otherlibs/dynlink/byte/dynlink.ml

```diff
@@ -95,46 +95,54 @@ module Bytecode = struct
 
   let run_shared_startup _ = ()
 
-  let run (ic, file_name, file_digest) ~unit_header ~priv =
+  let with_lock lock f =
+    Mutex.lock lock;
+    Fun.protect f
+      ~finally:(fun () -> Mutex.unlock lock)
+
+  let run lock (ic, file_name, file_digest) ~unit_header ~priv =
     let open Misc in
-    let old_state = Symtable.current_state () in
-    let compunit : Cmo_format.compilation_unit = unit_header in
-    seek_in ic compunit.cu_pos;
-    let code_size = compunit.cu_codesize + 8 in
-    let code = LongString.create code_size in
-    LongString.input_bytes_into code ic compunit.cu_codesize;
-    LongString.set code compunit.cu_codesize (Char.chr Opcodes.opRETURN);
-    LongString.blit_string "\000\000\000\001\000\000\000" 0
-      code (compunit.cu_codesize + 1) 7;
-    begin try DC.with_lock (fun () ->
-      Symtable.patch_object code compunit.cu_reloc;
-      Symtable.check_global_initialized compunit.cu_reloc;
-      Symtable.update_global_table ()
+    let clos = with_lock lock (fun () ->
```

Comment:
Can you comment on why you decided to widen the critical section here? Did we miss a protected-state access in the non-locked sections previously, or is it for some other reason?

Reply:
The section was widened to protect
```ocaml
        let old_state = Symtable.current_state () in
        ...
        if priv then Symtable.hide_additions old_state
```
from a concurrent write on the `Symtable`. And also because it seemed simpler to be as coarse as possible with the lock.